### PR TITLE
Bin borders for random threshold generation

### PR DESCRIPTION
**Pull request by inteldimitrius**
# Description
Utilizes bin borders for proper uniform random generation for thresholds.

### CODE REVIEW

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_hist_impl_dpc.cpp [hunk 1]
```diff
@@ -233,6 +233,37 @@ void train_kernel_hist_impl<Float, Bin, Index, Task>::init_params(train_context_
...
+    // Copy bin borders to the device
+    const Index max_bins = desc.get_max_bins();
+    bin_borders_device_ =
+        pr::ndarray<Float, 1>::empty(queue_, { ctx.column_count_ * max_bins }, alloc::device);
+
+    auto bin_borders_ptr = bin_borders_device_.get_mutable_data();
+    const sycl::range<1> range = de::integral_cast<std::size_t>(ctx.column_count_);
+    // Copying to the device is not convenient,
+    // because original bin borders are std::vector<ndarray<Float, 1>>
+    auto bin_borders_ptrs_host = pr::ndarray<const Float*, 1>::empty(queue_, { ctx.column_count_ });
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_hist_impl_dpc.cpp:245]:
```suggestion
    auto bin_borders_ptrs_host = pr::ndarray<Float*, 1>::empty(queue_, { ctx.column_count_ });
```

1) You'll change values here
2) Generally speaking shared memory usage is not recommended

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_hist_impl_dpc.cpp [hunk 2]
```diff
@@ -233,6 +233,37 @@ void train_kernel_hist_impl<Float, Bin, Index, Task>::init_params(train_context_
...
+    auto bin_borders_tmp_ptr = bin_borders_ptrs_host.get_mutable_data();
+    auto bin_borders_ptrs =
+        pr::ndarray<const Float*, 1>::empty(queue_, { ctx.column_count_ }, alloc::device);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_hist_impl_dpc.cpp:248]:
```suggestion
        pr::ndarray<Float*, 1>::empty(queue_, { ctx.column_count_ }, alloc::device);
```

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_hist_impl_dpc.cpp [hunk 3]
```diff
@@ -233,6 +233,37 @@ void train_kernel_hist_impl<Float, Bin, Index, Task>::init_params(train_context_
...
+    for (Index col_idx = 0; col_idx < ctx.column_count_; ++col_idx) {
+        bin_borders_tmp_ptr[col_idx] = ind_ftrs.get_bin_borders(col_idx).get_data();
+    }
+    bin_borders_ptrs.assign_from_host(queue_, bin_borders_tmp_ptr, ctx.column_count_)
+        .wait_and_throw();
+    auto origin_bin_borders_ptr = bin_borders_ptrs.get_data();
+    queue_
+        .submit([&](sycl::handler& cgh) {
+            cgh.parallel_for(range, [=](sycl::id<1> item) {
+                const Index col_idx = item;
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_hist_impl_dpc.cpp:258]:
```suggestion
    auto host_borders_event = bin_borders_ptrs.assign_from_host(queue_, bin_borders_tmp_ptr, ctx.column_count_);
    auto origin_bin_borders_ptr = bin_borders_ptrs.get_data();
    queue_
        .submit([&](sycl::handler& cgh) {
            cgh.depends_on(host_borders_event);
```

Please use events to achieve better concurrency

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_splitter_impl_dpc.cpp [hunk 4]
```diff
@@ -701,9 +704,15 @@ train_splitter_impl<Float, Bin, Index, Task, use_private_mem>::compute_random_sp
...
-                    static_cast<Index>((max_bin - min_bin + 1) * random_val);
-                Index ts_ftr_bin = min_bin + random_bin_ofs;
+                const Float* cur_ftr_bin_borders = bin_borders_ptr + ts_ftr_id * col_count;
+                const Float min_val = cur_ftr_bin_borders[min_bin];
+                const Float max_val = cur_ftr_bin_borders[max_bin];
+                const Float scaled_random_val = min_val + (max_val - min_val) * random_val;
+
+                Index cbin = min_bin;
+                for (; cbin <= max_bin && cur_ftr_bin_borders[cbin] < scaled_random_val; cbin++)
+                    ;
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_splitter_impl_dpc.cpp:714]:
```suggestion
                for (; cbin <= max_bin && cur_ftr_bin_borders[cbin] < scaled_random_val; ++cbin)
                    ;
```

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_splitter_sp_opt_impl_dpc.cpp [hunk 5]
```diff
@@ -203,10 +205,19 @@ sycl::event train_splitter_sp_opt_impl<Float, Bin, Index, Task, sbg_size>::rando
...
-                        const Index random_bin_ofs =
-                            static_cast<Index>(rand_val * (max_bin - min_bin + 1));
-                        ts.ftr_bin = min_bin + random_bin_ofs;
+                        const Float min_val = cur_ftr_bin_borders[min_bin];
+                        const Float max_val = cur_ftr_bin_borders[max_bin];
+                        const Float scaled_rand_val = min_val + (max_val - min_val) * rand_val;
+
+                        Index cbin = min_bin;
+                        for (; cbin <= max_bin && scaled_rand_val < cur_ftr_bin_borders[cbin];
+                             cbin++)
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_splitter_sp_opt_impl_dpc.cpp:217]:
```suggestion
                             ++cbin)
```

**Comment by icfaust** [comment]
@inteldimitrius Why was this PR closed?

**Pull request closed without merging**